packages/common/src/adapters/notification.ts

```diff
@@ -0,0 +1,500 @@
+
+function formatBaseNotification(notification: full.Notification) {
+  const timestamp = notification.actions[0].timestamp
+  return {
+    groupId: notification.groupId,
+    timestamp,
+    isViewed: !!notification.seenAt,
+    id: `timestamp:${timestamp}:group_id:${notification.groupId}`
+  }
+}
+
+function getDiscoveryEntityType({
+  type,
+  is_album
+}: {
+  type: string
+  is_album?: boolean
+}) {
+  if (type === 'track') {
+    return Entity.Track
+  } else if (is_album === true) {
+    return Entity.Album
+  }
+  return Entity.Playlist
+}
+
+/**
+ * Maps the SDK notifications type to the type that the UI expects,
+ * decoding hashIds and in some cases extracting userId from the groupId
+ * and other nuanced things on a per notification basis.
+ */
+export const notificationFromSDK = (
+  notification: full.Notification
+): Notification => {
+  if (notification.type === 'follow') {
+    const userIds = notification.actions
+      .map((action) => {
+        const data = action.data
+        return decodeHashId(data.followerUserId)
+      })
+      .filter(removeNullable)
+    return {
+      type: NotificationType.Follow,
+      userIds,
+      ...formatBaseNotification(notification)
+    }
+  } else if (notification.type === 'repost') {
+    let entityId = 0
+    let entityType = Entity.Track
+    const userIds = notification.actions
+      .map((action) => {
+        const data = action.data
+        entityId = decodeHashId(data.repostItemId) as number
+        entityType = getDiscoveryEntityType(data)
+        return decodeHashId(data.userId)
+      })
+      .filter(removeNullable)
+    return {
+      type: NotificationType.Repost,
+      userIds,
+      entityId,
+      entityType,
+      ...formatBaseNotification(notification)
+    }
+  } else if (notification.type === 'save') {
+    let entityId = 0
+    let entityType = Entity.Track
+    const userIds = notification.actions
+      .map((action) => {
+        const data = action.data
+        entityId = decodeHashId(data.saveItemId) as number
+        entityType = getDiscoveryEntityType(data)
+        return decodeHashId(data.userId)
+      })
+      .filter(removeNullable)
+    return {
+      type: NotificationType.Favorite,
+      userIds,
+      entityId,
+      entityType,
+      ...formatBaseNotification(notification)
+    }
+  } else if (notification.type === 'tip_send') {
+    const data = notification.actions[0].data
+    const amount = data.amount
+    const receiverUserId = decodeHashId(data.receiverUserId) as number
+    return {
+      type: NotificationType.TipSend,
+      entityId: receiverUserId,
+      entityType: Entity.User,
+      amount: amount!.toString() as StringWei,
+      ...formatBaseNotification(notification)
+    }
+  } else if (notification.type === 'tip_receive') {
+    const data = notification.actions[0].data
+    const amount = data.amount
+    const senderUserId = decodeHashId(data.senderUserId) as number
```

Comment:
Would `HashId.parse()` be the preferred thing here? can find/replace these

Reply:
Switched to `HashId.parse()` - lmk if I should go back to maintain parity but I think this is fine

Reply:
`HashId.parse` is fine but keep in mind it will throw for invalid input. The behavior in `ResponseAdapter` that I've been emulating is that for required id fields, we attempt to use `decodeHashId` to parse it and return `undefined` from the adapter if we can't parse the id field. Then it gets filtered out of the list later with `removeNullable`.

I don't _love_ that pattern. Just brining it up because throwing here will fail the whole request/saga if we aren't handling the error and just skipping the item.

Reply:
well in the components layer if userId is null/undefined the component returns null and we get empty `<li>`s, so if I went back to what it was before I think that's roughly the same behavior?

Though in practice we really _shouldn't_ be getting back any numbers that aren't hashable!!